torchao/prototype/low_bit_optim/subclass_4bit.py

```diff
@@ -181,3 +183,11 @@ def _(func, types, args, kwargs):
 
     # assume tensors from all ranks have the same signedness
     return OptimState4bit(codes, scale, x.qmap.clone(), x.signed, shape)
+
+
+# required by torch.distributed.checkpoint.save
+# note that we don't actually implement pin memory for this tensor subclass
+# (pin_memory argument is ignored in aten._to_copy)
+@OptimState4bit.implements(aten.is_pinned.default)
+def _(func, types, args, kwargs):
```

Comment:
@OptimState4bit.implements(aten.detach.default)
def _(func, types, args, kwargs):
    x = args[0]
    codes = x.codes.detach()
    scale = x.scale.detach()
    qmap = x.qmap.detach()
    return OptimState4bit(codes, scale, qmap, x.signed, x.shape)
    
    
    i think we would also need detach and slice

Reply:
I have added implementation for `aten.detach` in the code block above this. For `aten.slice`, I could not reproduce the case when this is needed. Will need reproduction steps from you.

Reply:
`aten.slice` is only called when world size changes. I have implemented this op for the tensor subclasses. Can you test this again? Also, torchtitan does not seem to support resume with different world size because stateful dataloader does not support it. So I'm guessing your problem with `aten.slice` did not come from torchtitan?

Reply:
@nighting0le01 Also lmk if you still face errors/problems with the current state of this PR, so we can proceed to merge it.

Reply:
@gau-nernst  LGTM, just tested! thanks